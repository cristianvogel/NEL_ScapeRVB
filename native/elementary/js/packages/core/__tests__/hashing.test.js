import {
  el,
  renderWithDelegate,
} from '..';


const InstructionTypes = {
  CREATE_NODE: 0,
  DELETE_NODE: 1,
  APPEND_CHILD: 2,
  SET_PROPERTY: 3,
  ACTIVATE_ROOTS: 4,
  COMMIT_UPDATES: 5,
};

class HashlessRenderer {
  constructor(config) {
    this.nodeMap = new Map();
    this.memoMap = new Map();
    this.batch = [];
    this.roots = [];

    this.nextMaskId = 0;
    this.maskTable = new Map();
  }

  getMaskId(hash) {
    if (!this.maskTable.has(hash)) {
      this.maskTable.set(hash, this.nextMaskId++);
    }

    return this.maskTable.get(hash);
  }

  getNodeMap() {
    return this.nodeMap;
  }

  getMemoMap() {
    return this.memoMap;
  }

  getRenderContext() {
    return { sampleRate: 44100 };
  }

  getActiveRoots() {
    return this.roots;
  }

  getTerminalGeneration() {
    return 4;
  }

  getBatch() {
    return this.batch;
  }

  clearBatch() {
    this.batch = [];
  }

  createNode(hash, type) {
    this.batch.push([InstructionTypes.CREATE_NODE, this.getMaskId(hash), type]);
  }

  deleteNode(hash) {
    this.batch.push([InstructionTypes.DELETE_NODE, this.getMaskId(hash)]);
  }

  appendChild(parentHash, childHash) {
    this.batch.push([InstructionTypes.APPEND_CHILD, this.getMaskId(parentHash), this.getMaskId(childHash)]);
  }

  setProperty(hash, key, val) {
    this.batch.push([InstructionTypes.SET_PROPERTY, this.getMaskId(hash), key, val]);
  }

  activateRoots(roots) {
    this.batch.push([InstructionTypes.ACTIVATE_ROOTS, roots.map(x => this.getMaskId(x))]);
  }

  commitUpdates() {
    this.batch.push([InstructionTypes.COMMIT_UPDATES]);
  }

  render(...args) {
    this.roots = renderWithDelegate(this, args);
  }
}

function sortInstructionBatch(x) {
  let copy = [...x];

  return copy.sort((a, b) => {
    if (a[0] === b[0]) {
      return a[1] < b[1] ? -1 : 1;
    }

    return a[0] < b[0] ? -1 : 1;
  });
}

// To test that our algorithm works even if the hashing function changes. We
// want to see that the same sets of instructions come through
test('instruction set similarity without hash values', function() {
  let tr = new HashlessRenderer();

  tr.render(el.cycle(440));
  expect(sortInstructionBatch(tr.getBatch())).toMatchSnapshot();
});

test('instruction set similarity without hash values 2', function() {
  let tr = new HashlessRenderer();

  let synthVoice = (hz) => el.mul(
    0.25,
    el.add(
      el.blepsaw(el.mul(hz, 1.001)),
      el.blepsquare(el.mul(hz, 0.994)),
      el.blepsquare(el.mul(hz, 0.501)),
      el.blepsaw(el.mul(hz, 0.496)),
    ),
  );

  let train = el.train(4.8);
  let arp = [0, 4, 7, 11, 12, 11, 7, 4].map(x => 261.63 * 0.5 * Math.pow(2, x / 12));

  let modulate = (x, rate, amt) => el.add(x, el.mul(amt, el.cycle(rate)));
  let env = el.adsr(0.01, 0.5, 0, 0.4, train);
  let filt = (x) => el.lowpass(
    el.add(40, el.mul(modulate(1840, 0.05, 1800), env)),
    1,
    x
  );

  let out = el.mul(0.25, filt(synthVoice(el.seq({seq: arp, hold: true}, train, 0))));
  tr.render(out, out);
  expect(sortInstructionBatch(tr.getBatch())).toMatchSnapshot();
});
