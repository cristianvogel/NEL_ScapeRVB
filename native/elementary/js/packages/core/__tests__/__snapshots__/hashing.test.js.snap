// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`instruction set similarity without hash values 1`] = `
[
  [
    0,
    0,
    "root",
  ],
  [
    0,
    1,
    "sin",
  ],
  [
    0,
    2,
    "mul",
  ],
  [
    0,
    3,
    "const",
  ],
  [
    0,
    4,
    "phasor",
  ],
  [
    0,
    5,
    "const",
  ],
  [
    2,
    0,
    1,
  ],
  [
    2,
    1,
    2,
  ],
  [
    2,
    2,
    3,
  ],
  [
    2,
    2,
    4,
  ],
  [
    2,
    4,
    5,
  ],
  [
    3,
    0,
    "channel",
    0,
  ],
  [
    3,
    3,
    "value",
    6.283185307179586,
  ],
  [
    3,
    5,
    "value",
    440,
  ],
  [
    4,
    [
      0,
    ],
  ],
  [
    5,
  ],
]
`;

exports[`instruction set similarity without hash values 2 1`] = `
[
  [
    0,
    0,
    "root",
  ],
  [
    0,
    1,
    "mul",
  ],
  [
    0,
    2,
    "const",
  ],
  [
    0,
    3,
    "svf",
  ],
  [
    0,
    4,
    "add",
  ],
  [
    0,
    5,
    "const",
  ],
  [
    0,
    6,
    "mul",
  ],
  [
    0,
    7,
    "const",
  ],
  [
    0,
    8,
    "mul",
  ],
  [
    0,
    9,
    "add",
  ],
  [
    0,
    10,
    "pole",
  ],
  [
    0,
    11,
    "const",
  ],
  [
    0,
    12,
    "mul",
  ],
  [
    0,
    13,
    "const",
  ],
  [
    0,
    14,
    "sin",
  ],
  [
    0,
    15,
    "mul",
  ],
  [
    0,
    16,
    "const",
  ],
  [
    0,
    17,
    "phasor",
  ],
  [
    0,
    18,
    "const",
  ],
  [
    0,
    19,
    "exp",
  ],
  [
    0,
    20,
    "mul",
  ],
  [
    0,
    21,
    "div",
  ],
  [
    0,
    22,
    "const",
  ],
  [
    0,
    23,
    "mul",
  ],
  [
    0,
    24,
    "div",
  ],
  [
    0,
    25,
    "sr",
  ],
  [
    0,
    26,
    "add",
  ],
  [
    0,
    27,
    "const",
  ],
  [
    0,
    28,
    "mul",
  ],
  [
    0,
    29,
    "mul",
  ],
  [
    0,
    30,
    "le",
  ],
  [
    0,
    31,
    "add",
  ],
  [
    0,
    32,
    "phasor",
  ],
  [
    0,
    33,
    "const",
  ],
  [
    0,
    34,
    "const",
  ],
  [
    0,
    35,
    "mul",
  ],
  [
    0,
    36,
    "mul",
  ],
  [
    0,
    37,
    "le",
  ],
  [
    0,
    38,
    "const",
  ],
  [
    0,
    39,
    "counter",
  ],
  [
    0,
    40,
    "mul",
  ],
  [
    0,
    41,
    "sub",
  ],
  [
    0,
    42,
    "sub",
  ],
  [
    0,
    43,
    "const",
  ],
  [
    0,
    44,
    "sub",
  ],
  [
    0,
    45,
    "add",
  ],
  [
    0,
    46,
    "mul",
  ],
  [
    0,
    47,
    "mul",
  ],
  [
    0,
    48,
    "add",
  ],
  [
    0,
    49,
    "mul",
  ],
  [
    0,
    50,
    "mul",
  ],
  [
    0,
    51,
    "const",
  ],
  [
    0,
    52,
    "add",
  ],
  [
    0,
    53,
    "blepsaw",
  ],
  [
    0,
    54,
    "blepsquare",
  ],
  [
    0,
    55,
    "blepsquare",
  ],
  [
    0,
    56,
    "blepsaw",
  ],
  [
    0,
    57,
    "mul",
  ],
  [
    0,
    58,
    "seq",
  ],
  [
    0,
    59,
    "const",
  ],
  [
    0,
    60,
    "mul",
  ],
  [
    0,
    61,
    "const",
  ],
  [
    0,
    62,
    "mul",
  ],
  [
    0,
    63,
    "const",
  ],
  [
    0,
    64,
    "mul",
  ],
  [
    0,
    65,
    "const",
  ],
  [
    0,
    66,
    "root",
  ],
  [
    2,
    0,
    1,
  ],
  [
    2,
    1,
    2,
  ],
  [
    2,
    1,
    3,
  ],
  [
    2,
    3,
    4,
  ],
  [
    2,
    3,
    5,
  ],
  [
    2,
    3,
    6,
  ],
  [
    2,
    4,
    7,
  ],
  [
    2,
    4,
    8,
  ],
  [
    2,
    6,
    2,
  ],
  [
    2,
    6,
    52,
  ],
  [
    2,
    8,
    9,
  ],
  [
    2,
    8,
    10,
  ],
  [
    2,
    9,
    11,
  ],
  [
    2,
    9,
    12,
  ],
  [
    2,
    10,
    19,
  ],
  [
    2,
    10,
    20,
  ],
  [
    2,
    12,
    13,
  ],
  [
    2,
    12,
    14,
  ],
  [
    2,
    14,
    15,
  ],
  [
    2,
    15,
    16,
  ],
  [
    2,
    15,
    17,
  ],
  [
    2,
    17,
    18,
  ],
  [
    2,
    19,
    21,
  ],
  [
    2,
    20,
    44,
  ],
  [
    2,
    20,
    45,
  ],
  [
    2,
    21,
    22,
  ],
  [
    2,
    21,
    23,
  ],
  [
    2,
    23,
    24,
  ],
  [
    2,
    23,
    25,
  ],
  [
    2,
    24,
    26,
  ],
  [
    2,
    24,
    27,
  ],
  [
    2,
    26,
    28,
  ],
  [
    2,
    26,
    29,
  ],
  [
    2,
    28,
    30,
  ],
  [
    2,
    28,
    31,
  ],
  [
    2,
    29,
    42,
  ],
  [
    2,
    29,
    43,
  ],
  [
    2,
    30,
    32,
  ],
  [
    2,
    30,
    33,
  ],
  [
    2,
    31,
    35,
  ],
  [
    2,
    31,
    36,
  ],
  [
    2,
    32,
    34,
  ],
  [
    2,
    35,
    37,
  ],
  [
    2,
    35,
    38,
  ],
  [
    2,
    36,
    41,
  ],
  [
    2,
    36,
    33,
  ],
  [
    2,
    37,
    39,
  ],
  [
    2,
    37,
    40,
  ],
  [
    2,
    39,
    30,
  ],
  [
    2,
    40,
    38,
  ],
  [
    2,
    40,
    25,
  ],
  [
    2,
    41,
    5,
  ],
  [
    2,
    41,
    37,
  ],
  [
    2,
    42,
    5,
  ],
  [
    2,
    42,
    30,
  ],
  [
    2,
    44,
    5,
  ],
  [
    2,
    44,
    19,
  ],
  [
    2,
    45,
    46,
  ],
  [
    2,
    45,
    47,
  ],
  [
    2,
    46,
    30,
  ],
  [
    2,
    46,
    48,
  ],
  [
    2,
    47,
    42,
  ],
  [
    2,
    47,
    51,
  ],
  [
    2,
    48,
    49,
  ],
  [
    2,
    48,
    50,
  ],
  [
    2,
    49,
    37,
  ],
  [
    2,
    49,
    5,
  ],
  [
    2,
    50,
    41,
  ],
  [
    2,
    50,
    51,
  ],
  [
    2,
    52,
    53,
  ],
  [
    2,
    52,
    54,
  ],
  [
    2,
    52,
    55,
  ],
  [
    2,
    52,
    56,
  ],
  [
    2,
    53,
    57,
  ],
  [
    2,
    54,
    60,
  ],
  [
    2,
    55,
    62,
  ],
  [
    2,
    56,
    64,
  ],
  [
    2,
    57,
    58,
  ],
  [
    2,
    57,
    59,
  ],
  [
    2,
    58,
    30,
  ],
  [
    2,
    58,
    51,
  ],
  [
    2,
    60,
    58,
  ],
  [
    2,
    60,
    61,
  ],
  [
    2,
    62,
    58,
  ],
  [
    2,
    62,
    63,
  ],
  [
    2,
    64,
    58,
  ],
  [
    2,
    64,
    65,
  ],
  [
    2,
    66,
    1,
  ],
  [
    3,
    0,
    "channel",
    0,
  ],
  [
    3,
    2,
    "value",
    0.25,
  ],
  [
    3,
    3,
    "mode",
    "lowpass",
  ],
  [
    3,
    5,
    "value",
    1,
  ],
  [
    3,
    7,
    "value",
    40,
  ],
  [
    3,
    11,
    "value",
    1840,
  ],
  [
    3,
    13,
    "value",
    1800,
  ],
  [
    3,
    16,
    "value",
    6.283185307179586,
  ],
  [
    3,
    18,
    "value",
    0.05,
  ],
  [
    3,
    22,
    "value",
    -1,
  ],
  [
    3,
    27,
    "value",
    6.91,
  ],
  [
    3,
    33,
    "value",
    0.5,
  ],
  [
    3,
    34,
    "value",
    4.8,
  ],
  [
    3,
    38,
    "value",
    0.01,
  ],
  [
    3,
    43,
    "value",
    0.4,
  ],
  [
    3,
    51,
    "value",
    0,
  ],
  [
    3,
    58,
    "seq",
    [
      130.815,
      164.81657214199782,
      196.0010402616231,
      246.94583642691148,
      261.63,
      246.94583642691148,
      196.0010402616231,
      164.81657214199782,
    ],
  ],
  [
    3,
    58,
    "hold",
    true,
  ],
  [
    3,
    59,
    "value",
    1.001,
  ],
  [
    3,
    61,
    "value",
    0.994,
  ],
  [
    3,
    63,
    "value",
    0.501,
  ],
  [
    3,
    65,
    "value",
    0.496,
  ],
  [
    3,
    66,
    "channel",
    1,
  ],
  [
    4,
    [
      0,
      66,
    ],
  ],
  [
    5,
  ],
]
`;
